### PR TITLE
Implement breadthFirstSearch and MaximalIndependentSet from PBBS

### PR DESCRIPTION
**Pull request by MagicfTail**
The test files are hosted on a temporary erda server as they are pretty large. The links in external-data.txt should be changed to the official erda archive later.  The test files have been generated using the same parameters as the PBBS files.

Benchmarks for the MIS and BFS algorithms have been run with both a multicore and cuda backend, and have been compared to the PBBS implementations.

```
-------------------------------MIS------------------------------
Futhark-cuda:
    randLocalGraph_J_10_20000000    :   0.322635
    rMatGraph_J_12_16000000         :   0.425259
    3Dgrid_J_64000000               :   1.2356

Futhark-multicore:
    randLocalGraph_J_10_20000000    :   4.79
    rMatGraph_J_12_16000000         :   5.1885
    3Dgrid_J_64000000               :   13.0593

PBBS-incrementalMIS:
    randLocalGraph_J_10_20000000    :   2.011
    rMatGraph_J_12_16000000         :   1.149
    3Dgrid_J_64000000               :   2.598

PBBS-ndMIS:
    randLocalGraph_J_10_20000000    :   1.528
    rMatGraph_J_12_16000000         :   2.251
    3Dgrid_J_64000000               :   2.147


------------------------------BFS------------------------------
Futhark-cuda:
    randLocalGraph_J_10_20000000    :   0.240602
    rMatGraph_J_12_16000000         :   0.291458
    3Dgrid_J_64000000               :   0.963533

Futhark-multicore:
    randLocalGraph_J_10_20000000    :   5.3723
    rMatGraph_J_12_16000000         :   5.7308
    3Dgrid_J_64000000               :  34.4412

PBBS-simple:
    randLocalGraph_J_10_20000000    :   1.932
    rMatGraph_J_12_16000000         :   2.049
    3Dgrid_J_64000000               :  16.179

PBBS-backForwardBFS:
    randLocalGraph_J_10_20000000    :   0.821
    rMatGraph_J_12_16000000         :   0.701
    3Dgrid_J_64000000               :  16.939

PBBS-deterministicBFS:
    randLocalGraph_J_10_20000000    :   2.649
    rMatGraph_J_12_16000000         :   2.578
    3Dgrid_J_64000000               :  27.358
```

The benchmarks have been run on slurm clusters using the following paramters
```
#SBATCH -p gpu --gres=gpu:titanrtx
#SBATCH --ntasks=1
#SBATCH --cpus-per-task=4
#SBATCH --mem=64000M
```

### CODE REVIEW

File: pbbs/breadthFirstSearch/breadthFirstSearch.fut [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    else {vertex = -1, parent = -1}
+
+def remove_duplicates_GPU (queue: []queuePair) = 
+    -- We need to append an extra in case the queue only contains items with the same vertex value
+    -- Array copy is hopefully optimized away by the compiler
+    let sorted = radix_sort_int_by_key (\q -> q.vertex) i32.num_bits i32.get_bit queue ++ [{vertex = -1, parent = -1}]
+    in  map2 get_first sorted (rotate (1) sorted)
+
+-- Unused
+-- Might also not be secure as the function "(\_ b -> b)" isn't commutative
```
Comment (athas) [inline pbbs/breadthFirstSearch/breadthFirstSearch.fut:18]: This is very dangerous.  You need to find another way of doing this.  What if you defined the operator to pick the element with the highest `vertex` field?

However, you should also note that this is a pretty slow operator you're using here.  It is going to require locking at run-time.

File: pbbs/breadthFirstSearch/breadthFirstSearch.fut [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+    in BFS vertexes_enc edges_enc parents queue
+
+-- ==
+-- input @ data/randLocalGraph_J_10_20000000.in
+-- output @ data/randLocalGraph_J_10_20000000.out
+-- input @ data/rMatGraph_J_12_16000000.in
+-- output @ data/rMatGraph_J_12_16000000.out
+-- input @ data/3Dgrid_J_64000000.in
+-- output @ data/3Dgrid_J_64000000.out
```
Comment (athas) [inline pbbs/breadthFirstSearch/breadthFirstSearch.fut:84]: Add missing newline.

File: pbbs/breadthFirstSearch/breadthFirstSearch.fut (hunk 1 above)
Comment (athas) [reply pbbs/breadthFirstSearch/breadthFirstSearch.fut:18]: (I understand you don't actually use this function, but then why keep it?)

File: pbbs/maximalIndependentSet/maximalIndependentSet.fut [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+    in MIS vertexes_enc edges_enc random_state C I indexes nEdges
+
+-- ==
+-- input @ data/randLocalGraph_J_10_20000000.in
+-- output @ data/randLocalGraph_J_10_20000000.out
+-- input @ data/rMatGraph_J_12_16000000.in
+-- output @ data/rMatGraph_J_12_16000000.out
+-- input @ data/3Dgrid_J_64000000.in
+-- output @ data/3Dgrid_J_64000000.out
```
Comment (athas) [inline pbbs/maximalIndependentSet/maximalIndependentSet.fut:88]: Add missing linebreak.

File: pbbs/breadthFirstSearch/breadthFirstSearch.fut (hunk 1 above)
Comment (MagicfTail) [reply pbbs/breadthFirstSearch/breadthFirstSearch.fut]: I left them in as I wasn't 100% sure that there was no way of making them safe, and they had some performance benefits over the safe method. They have now been removed

Comment (athas) [reply pbbs/breadthFirstSearch/breadthFirstSearch.fut]: Well, there might indeed be some tricks.  Sorting is not the fastest way to deduplicate.  What is the potential speedup?

**Comment by MagicfTail** [comment]
> I would like to see a comment somewhere on validation. How do we know these results are correct? fut2pbbs.c has not been modified, so we cannot convert the results back to PBBS and use their validator.

The output format required by PBBS is integer sequence format, which is already supported by fut2pbbs (sequenceInt), so it's already possible to test the output.

File: pbbs/breadthFirstSearch/breadthFirstSearch.fut (hunk 1 above)
Comment (MagicfTail) [reply pbbs/breadthFirstSearch/breadthFirstSearch.fut]: The potential speedup is around 4x if the best results are cherry picked from each of the different unsafe implementations. Some cases are however fastest with the current way of removing duplicates

Comment (MagicfTail) [reply pbbs/breadthFirstSearch/breadthFirstSearch.fut]: I have changed to filtering method to not rely on sorting, which resulted in an overall speedup of around 2x.

**Comment by athas** [comment]
BFS fails for me:

```
$ futhark bench --backend=cuda breadthFirstSearch.fut
Compiling breadthFirstSearch.fut...
Reporting arithmetic mean runtime of at least 10 runs for each dataset (min 0.5s).
More runs automatically performed for up to 300s to ensure accurate measurement.

breadthFirstSearch.fut (no tuning file):
data/randLocalGraph_J_10_20000000.in:
breadthFirstSearch.fut.actual and breadthFirstSearch.fut.expected do not match:
Value #0 index [1]: expected 19999998, got 8
...and 10575413 other mismatches.
data/rMatGraph_J_12_16000000.in:
breadthFirstSearch.fut.actual and breadthFirstSearch.fut.expected do not match:
Value #0 index [23]: expected 12582916, got 1
...and 7605390 other mismatches.
data/3Dgrid_J_64000000.in:
breadthFirstSearch.fut.actual and breadthFirstSearch.fut.expected do not match:
Value #0 index [200]: expected 201, got 199
...and 32240000 other mismatches.
```

**Comment by athas** [comment]
There is still no documentation on how to verify the results.  I think this works: I grab e.g. `randLocalGraph_J_10_20000000` from PBBS, convert the corresponding .out file from Futhark to PBBS format with `fut2pbbs`, then use `MISCheck` from PBBS.

**Comment by athas** [comment]
On `master` I have modified `fut2pbbs` so that it can convert the maximalIndependentSet Futhark input and output files back into the PBBS format.  MISCheck from PBBS can then be used.  It seems to work.

**Comment by athas** [comment]
Works now.  I need to take a quick look in a profiler one of the coming days, but then I expect to merge it.

**Pull request merged**